### DISCUSSION TITLE
Improve Bookmarks Management

### DISCUSSION THREAD

**Discussion by EmmanuelMiquet** [organicmaps/organicmaps#7975] [Q&A]
Hi all!

I'm a bit new here... after having used maps.me for ages, been disappointed with new direction and happy to find Organic Maps with "good ol' simple OSM based app"!
Now, I'm still using a day to day hybrid of Google Maps for places search, ratings... (and "My Maps", which is *this* close to be a "perfect app", but still misses some spots for me) and Organic Maps, for instance for travels and hikes (with more hiking paths, better walking path finding..., and slightly better bookmark management)
I somehow port bookmarks from one app to the other through KML etc.

But I still fail to find a proper bookmarks management interface on either side (or a third party for sync, but using 2 apps is already annoying to manage)

I started looking for opened points and ideas in this Github, but it seems some discussion area are more active than others and I'm a bit new to contribution to such open source projects.

From a "design" / UX perspective, I'd wish to see a traditional "tree-like" approach to bookmarks, where lists would be like folders, containing either sub-folders or leaf nodes (paths, landmarks)
Optionally there could be some additional features, like tags, multiple lines edits etc...
I read some parts of discussions on the matter and what would be the issues because this is mobile (and cross-platform) world, but mixing different existing applications look and feel, I think that something quite intuitive could be brought up. (or even propose switch between display modes if implemented)

The need for such tree-like approach could, first and quite obviously, bring a geography approach in bookmarks organizing (like Country -> region -> city)
Another option for user, for instance travelling, I really like to split between "historical" places, nature sightseeing, hikes, potential nice accommodations, restaurants... and being able to group them (but not having to look for them in hundreds of lists) and display only what I need at some point would be very (very) helpful.

I was wondering if this is something already planned, or in a backlog of some form, or if this kind of feature had been rejected for some reason.

If not, I'm quite newbie to mobile dev, but I'd be glad to take this as a learning opportunity and give it a go by small incremental steps (or with collaboration!) and try to deliver such feature if I can manage it (and if the need seems relevant)

Thanks in advance for answers / suggestions / recommendations for where to start...!

Edit: I'm adding 2 screenshots from apps with potential "look & feel" of such tree-structure, with either
* a very "desktop-like" approach in "MyLifeOrganized" with (possibly nested) folders and items, that can be flagged, starred... (and multi-selectable, editables, copiable...), there are a lot of sub-features there, but the "UX" is nice
![MLO](https://github.com/organicmaps/organicmaps/assets/16577680/8f7bb64d-e2e9-44f9-8f67-56eed93ec14b)

* a more "app like" / windows explorer approach in ZArchiver, where current path is displayed and navigation to parent is done with a ".." line, so might be more cumbersome to navigate but easier to "click", that maybe a good "mvp" target though, and a good basis for incremental addition of other sub-features 
![ZArchiver](https://github.com/organicmaps/organicmaps/assets/16577680/d9867ae6-435a-4cf8-ad5e-72ef6a463a3a)

**Comment by hb0nd** [comment]
a tree structure is a good idea :)

**Reply by biodranik** [reply]
Thanks for starting the discussion! I would suggest focusing on the root use cases before proceeding with the implementation suggestions.

1. When do you need a tree structure and why? At home for planning and bookmarking? Or on the run in the wild to quickly find something you need? Do you actually need a tree structure? Can it be done more efficiently/easier in a different way?
2. What is the value of organizing bookmarks by countries regions and cities? How would you use this mode? When planning? Or when traveling? It is already so easy to see bookmarks in a given country or region directly on the map. So what is the use-case? 
3. Will improving the "sort mode" help here? There are already "by distance" and "by type" sorting modes. Isn't the latter one that you need? Maybe it can be improved though, if yes, then how? What is the use-case again?
4. Can a better search help in your case? How?

It is important to distinguish between using app for planning the trip and using the app during the trip.

**Reply by EmmanuelMiquet** [reply]
Thanks for the input! It's indeed interesting to question then "why" and "what" before the "how" ^^

1. Good point! I think in both use cases, planning phase and "in the field". I'm using the app mostly for 2 situations: hikes and travel.

"tree" structure was not necessarily the exact appropriate name, it was more from a UI aspect, the main goal would be to avoid having a single huge top level "list of lists".

For the planning aspect, for instance, I tend to have really many lists, and inside possibly many bookmarks (either I split in smaller lists, but end up with even more lists, or bigger lists, but many bookmarks annoying to scroll through / partially select...)
But technically the _planning_ part is a "one time" annoying part, where it could probably mostly be done "offline", regularly grooming etc...

For "on the field" part however, that's probably the most "daily usage" important part.
Again having many bookmarks, maps areas can tend to be cluttered, overlapping etc... and scrolling through lists to check which are shown or hidden is not very practical
Being able to filter in a "layer" manner would be a great improvement.

However, maybe adding a "tag" mechanism (for search, filtering...) would do it ; but combining it with the "organizing" aspect of planning would be really cool! 

Maybe such "tag"/"type" approach already exists internally, inherent to the map info of the bookmark, like food, accommodation, park, nice viewpoint (and allowing tag addition to user may be "dangerous" ... like we can see on map icons and search, but when importing bookmarks for instance, this is not taken into account if there is a registered location at the same place... Also it's currently not possible to filter or select by this criteria in the bookmark if I'm correct.

2. The "by countries / regions / cities" may be a bad example, as it can be purely "geographically" seen on the map indeed. 

It could be beneficial for the sake of organizing in the planning phase. Again, maybe this is not the "developer intended usage", but I do tend to create lists geography based (like travel to X planned = boom new list, possibly with sub-lists for hikes, places to sleep/eat...), and end up with many (many) lists, and having everything on a single huge "list of lists" level quickly becomes unmanageable.

Also, the "on the go" case could be more covered by the "layering" aspect to avoid cluttering / overlapping info indeed, as long as there is a way to select / filter more intuitively across bookmarks.

3. To be honest, as a user, I did not even know there was a sorting mechanism until now ^^ the way I see it existing would be insufficient for my use cases as it's only to sort items inside a list, and only by distance around (or by type), but not by "tag" / layer.

4.  Could be, yes. Again, maybe a "tag" approach could be interesting (or complimentary), to mark landmarks and be able to filter by tag. Also, it could be very helpful to make "tag-wise" bookmarks styling (icon, color...)


I'd be happy to further discuss the options that you'd see fit for the app philosophy! Of course, this would need to be decomposed in smaller steps/stories.
I will try to get familiar with the repo, tools and local setup... in the meantime!

**Reply by biodranik** [reply]
Thanks for the detailed answer! Now it's clearer that we need to find elegant and simple solutions to a few key points:
1. A lot of lists or bookmarks become "unmanageable". Let's define this "unmanageable" and split it into more detailed use-cases (search? Adding new bookmarks? Modifying?)
2. How to simplify one-time planning/adding? How to make it easier on the mobile device? Maybe on the web?
3. How to simplify in-the-field use cases?

Most users don't like complex stuff, and even may not discover it (like you and the sorting). It would be great to check how other apps do it in a better way. Recently one user asked to apply the approach from gmail app :)

Doing things in a complex way is often easier than designing things in a simpler way.

We would love to have more involved contributors and maintainers, there are so many tasks and things to improve! :)

**Reply by EmmanuelMiquet** [reply]
Sorry for the very delayed answer, indeed, time is a scarce resource ^^ + had a bit of hard time setting up local environment and finally discover a bit the code base and desktop app...!

1. "Unmanageable" in several way from my user experience: scroll and scroll again through many lists, no "list name search" option etc. I see list of bookmarks in two ways: either as "layers" in the map, so visual clustering (area) and separation (color, type...) of bookmarks, maybe more for in-the-field usage, and more as an "organizer" (maybe more in a "planning" aspect)

So yes, maybe a dedicated search option in the bookmark widgets, or a dedicated "search in bookmark" option in search mode!

Also the fact that you can only show / hide a whole list, so this pushes to split lists to show cluttered info with more granularily, but leading to even more lists, so hard to find / select only few ones. So again, maybe the search approach would be a good step at least in the "declutering" aspect to only display a certain type of bookmarks, or by name search etc.

2. No really additional idea here... the point below on KML issues I found may solve part of this by keeping the "main / one-time" planning part outside of the app, and fixing some import subtelties. Again, the "tree" aspect with a "layering" aspect seem really appealing to me, but indeed, it would need more praise / approval, be very user friendly, and the transition / migration with current should be the smoothest possible, so no rush there ^^

3. Filter by type could definetly help, so that even if all bookmarks are mixed up in big lists (tedious to hide / show appropriate ones in the "big list of lists"), it would be easy to say "ok, do I have any cool bookmarked restaurant, bar, park... around?".
I see the "default" filtering option in the bookmark lists seem to order by feature type, but it doesn't allow to filter in the map (show only feature x of bookmark list)

So either through the "search" approach, or maybe allow for partial list "show/hide" based on type?


As mentioned above though:
I also realize one "side quest" that would partly delay the need for these aspects would be to improve / fix the KML import in some way, especially coming from Google's "MyMaps", which is the one I struggled using from a user perspective.
Indeed, if the KML import is enhanced, it would partly cover the "planning" phase, as it would allow to make plans in My Maps or other KML generating app, and simply sync / import it when ready.

The issue is that MyMaps exports KMLs with dummy icons not matching the ones selected in the map, and on OM side, it looks like the import ignores the color indication from these KMLs and set them to default red.
Plus, on such import, it looks like it ignores if there is a "feature type" at location of the bookmark and only uses default icon, unlike bookmarks created within the app.
So if that was fixed, it would at least help visually upon import amongst many (many) bookmarks, even with everything switched on.

In the meantime, I'm getting familiar with the repo and will maybe have a look at the "good first issues" to practice some dev in there :)

**Reply by biodranik** [reply]
Importing custom icons from Google is tricky, it requires:
1. Retrieving them from Google via https (may not be accessed without authorization).
2. Rendering custom png/jpg images in OM.

See #2795

**Reply by EmmanuelMiquet** [reply]
But without importing it (or as a default reaction if no valid icon is imported), couldn't the bookmark location be checked (with a very short distance threshold) with local OSM data and if there is something (not a custom coordinates point in the middle of nowhere), there should be the icon of the place that could be retrieved and used for the bookmark icon / feature type (for instance a park, a monument...)?

**Reply by biodranik** [reply]
All current OM bookmark icons are hard-coded. Any custom icons require implementing it, for example, by rendering SVGs (there was an issue about it), or by rendering externally imported png/jpg

**Comment by bad0c0de** [comment]
I also would like to be able to efficiently search within my bookmarked locations. I understand that there are at least two standard ways of organising items: either rely on a hierarchical structure (e.g. folders, with some level of recursivity) or rely on tags. 

I would personally love to see support assigning and searching for multiple tags in the case of bookmarks: I could the search for #Restaurants in #Toulouse or alternatively #Friends in #Zurich...
